### PR TITLE
Implement @_downgrade_exhaustivity_check

### PR DESCRIPTION
**Pull request by CodaFi**
Dispatch requests the ability to add a new case, but to treat missing
instances of that case in patterns as warnings instead of errors.  It is
still an error to make reference to the annotated case in at least one
pattern then not cover the rest of the space, but it is not an error
to omit the space of patterns referencing the case entirely.

This attribute is private and uglified to intentionally discourage
its use outside just this one use case.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by jrose-apple** [comment]
For Apple folks: this is rdar://problem/31100201.

File: test/attr/attr_downgrade_exhaustivity.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// RUN: %target-typecheck-verify-swift -swift-version 3 %s
+
+enum Runcible {
+  case spoon
+  case hat
+  @_downgrade_exhaustivity_check
```
Comment (jckarter) [inline test/attr/attr_downgrade_exhaustivity.swift]: How much harder would it be to make the attribute versioned, so we can make these into errors in Swift 4 mode?

File: test/attr/attr_downgrade_exhaustivity.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  case (.spoon, .spoon):
+    break
+  case (.spoon, .hat):
+    break
+  case (.hat, .spoon):
+    break
+  case (.hat, .hat):
+    break
+  }
+}
```
Comment (jckarter) [inline test/attr/attr_downgrade_exhaustivity.swift]: Should we raise an error if `fork` is nonexhaustively used in parts of the pattern, as in:

```
switch (x!, y!) {
case (.spoon, _):
  break
case (.hat, _):
  break
case (.fork, .spoon):
  break
case (.fork, .hat):
  break
}
```

Is it worth testing other patterns where the enum appears in structural position, like nested in `Optionals` or `Trioptional` declared above?

File: test/attr/attr_downgrade_exhaustivity.swift (hunk 1 above)
Comment (CodaFi) [inline test/attr/attr_downgrade_exhaustivity.swift]: Your comments made me realize I left out half of your test cases from before.  Let me add them back.

File: test/attr/attr_downgrade_exhaustivity.swift (hunk 2 above)
Comment (CodaFi) [inline test/attr/attr_downgrade_exhaustivity.swift]: For completeness: I do agree that this should happen the way you intended to, and this patch implements that.  Gonna push the right tests.

File: test/attr/attr_downgrade_exhaustivity.swift (hunk 1 above)
Comment (jckarter) [inline test/attr/attr_downgrade_exhaustivity.swift]: I don't think that's necessary, I just missed the fact it was already conditional on version. Thanks.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test and merge

**Comment by CodaFi** [comment]
[Merge happening here](https://ci.swift.org/view/Pull%20Request/job/swift-PR-smoke-merge/3003/)

**Comment by CodaFi** [comment]
⛵️ 

Will nominate for 4.0 when the overlay changes come in.

**Comment by jrose-apple** [comment]
I don't think we're bothering to pull this into 4.0. We'll rebranch soon.

**Comment by CodaFi** [comment]
Then I *would* like to see about getting #9931 in if possible.

**Pull request merged**